### DISCUSSION TITLE
ElkJS/Dagre with subflows?

### DISCUSSION THREAD

**Discussion by sylvainmarie** [xyflow/xyflow#3495] [Layouting] [answered]
Hi, 
When I use the Subflow example data with the Dagre or ElkJS examples, the resulting layout doesn't properly use the subflows (read: the result is weird 😅 ). The libraries are supposed to properly layout subflows, woudl it be some issue with their ReactFlow integration?

SubFlow example, manual layout:
![image](https://github.com/wbkd/react-flow/assets/4747407/6ebea7fd-bc67-409d-960c-0599cd4773a0)

ElkJs example, using the SubFlow sample data 😨:
![image](https://github.com/wbkd/react-flow/assets/4747407/1ecceb0e-3152-4ef2-a32b-2095a7064237)

**Comment by moklick** [comment, marked as answer]
Hey @sylvainmarie 

dagre doesn't support sub flows. If you are working with elk, you need to change the format for elk a bit. Elk doesn't work with a flat but a nested node structure, so you need to adjust the nodes array like this:

```js
const graph = {
  id: "root",
  layoutOptions: { 'elk.algorithm': 'layered' },
  children: [
    { id: "n1", width: 30, height: 30, children: [
       { id: "c1", width: 30, height: 30 },
       { id: "c2", width: 30, height: 30 },
    ]},
    { id: "n2", width: 30, height: 30 },
    { id: "n3", width: 30, height: 30 }
  ],
  edges: [
    { id: "e1", sources: [ "n1" ], targets: [ "n2" ] },
    { id: "e2", sources: [ "n1" ], targets: [ "n3" ] }
  ]
};
```

As you can see, elk doesn't use the `parentNode` option but expects a `children` array for the child nodes.

**Reply by sylvainmarie** [reply]
Oh, I see. Time to write the flat<->hierarchical encoding/decoding bit of code. Thanks for the quick answer!

**Reply by zeyupan888** [reply]
```
    const ELK = require('elkjs')
    const elk = new ELK()

    elk.layout(graph)
        .then(console.log)
        .catch(console.error)
```
Hello, may I know after elk.layout, what should we do? Thanks.

**Comment by zeyupan888** [comment]
I got wired layout using elk
```

    const graph = {
        id: "root",
        layoutOptions: { 'elk.algorithm': 'layered' },
        children: [
            {
                id: "n1", width: 30, height: 30, children: [
                    { id: "c1", width: 30, height: 30 },
                    { id: "c2", width: 30, height: 30 },
                ]
            },
            { id: "n2", width: 30, height: 30 },
            { id: "n3", width: 30, height: 30 }
        ],
        edges: [
            { id: "e1", sources: ["n1"], targets: ["n2"] },
            { id: "e2", sources: ["n1"], targets: ["n3"] }
        ]
    };

    const ELK = require('elkjs')
    const elk = new ELK()

    const [nodes, setNodes] = useState<Node[]>([]);
    const [edges, setEdges] = useState<Edge[]>([]);


    const layoutNodes = (node: any) => {
        const newNode = {
            id: node.id,
            position: { x: node.x, y: node.y },
            type: 'default',
            data: { label: node.id },
        };

        let elements = [newNode];

        if (node.children) {
            node.children.forEach((childNode: any) => {
                elements = elements.concat(layoutNodes(childNode));
            });
        }

        return elements;
    };

    useEffect(() => {
        elk.layout(graph)
            .then((layout: any) => {
                const newNodes = layoutNodes(layout);
                const newEdges = layout.edges.map((edge: any) => ({
                    id: edge.id,
                    source: edge.sources[0],
                    target: edge.targets[0],
                    type: 'default',
                }));

                setNodes(newNodes);
                setEdges(newEdges);
            })
            .catch(console.error);
    }, []);


    console.log('nodes is', nodes)
    console.log('edges are', edges)

    return (
        <div style={{ width: '100%', height: '100vh' }}>
            <ReactFlow
                nodes={nodes}
                edges={edges}
                nodesDraggable={false}
                nodesConnectable={false}
                nodeTypes={undefined}
            />
        </div>
    );
```

<img width="409" alt="image" src="https://github.com/wbkd/react-flow/assets/71034205/f75c57dd-b7a4-40da-97b4-9b5dd2797c7e">


please help out, thanks!

**Reply by aaka5h** [reply]
only thing you are missing out is applying the width of nodes calculated by elk
```js
const newNode = {
// ...
style = { width: node.width, height: node.height }
// ...
}
```

refer to this complete example:
https://codesandbox.io/p/sandbox/elkjs-layout-subflows-9og9hl

**Comment by zeyupan888** [comment]
hmmm, finally figured it out, thank you all!
Any advice on the style?
<img width="391" alt="image" src="https://github.com/wbkd/react-flow/assets/71034205/f91dae51-343a-48c7-beba-ba731c649ca9">

1. the line that splits from n1 is wired, partial of the line is overlapping with another node. 
2. is there a better way to do arrow edge instead of edge marker?
Thank you very much

**Reply by debasis9658** [reply]
Hi @zeyupan888 Can you please add the code how did you achieve this?

**Reply by zeyupan888** [reply]
hiii, you can do it with elkjs children.
https://rtsys.informatik.uni-kiel.de/elklive/

**Reply by zeyupan888** [reply]
together with reactflow
https://reactflow.dev/examples/layout/elkjs

**Reply by satya-nutella** [reply]
@zeyupan888 how did you get the parent nodes of the subflow to be of the right height. I am having issues with that other than that the subflow layout works. Any help would be awesome

**Reply by SumitSaxenaTec** [reply]
Hi @zeyupan888 Can you please add the code or example how did you achieve this?

**Reply by sylvainmarie** [reply]
> Hi @zeyupan888 Can you please add the code or example how did you achieve this?

Hi, have a look at https://github.com/sylvainmarie/react-flow-test/blob/d41f0a2f73892a0e8738de21c3afe570e10deecd/src/app/components/ReactFlowContainer.tsx#L38 to see the data conversion between reactflow and elkjs (tree <-> flat) to layout subgraphs
